src/mococrw/padding_mode.h

```diff
@@ -25,116 +25,87 @@
+     * @brief Check if a message of a given size can be encrypted using the given padding.
+     * @param key The RSA public key that will be used for encryption
+     * @param messageSize The size of the message that will be encrypted
      * @return the maximum size of the data that can be encrypted in bytes.
      */
-    int getDataBlockSize(const AsymmetricPublicKey &key) const;
+    virtual bool checkMessageSize(const AsymmetricPublicKey &key, size_t messageSize) const = 0;
 
     /**
      * @brief Prepares the given openssl context with the padding specific parameters.
      * @param ctx OpenSSL PKEY context
      */
-    void prepareOpenSSLContext(openssl::SSL_EVP_PKEY_CTX_Ptr& ctx) const;
+    virtual void prepareOpenSSLContext(openssl::SSL_EVP_PKEY_CTX_Ptr& ctx) const = 0;
 };
 
-
 /**
- * @brief PKCSEncryptionPadding
- *
- * This class defines the parameters specific to the PKCS1 padding mode.
+ *  @brief RSASignaturePadding
  *
+ *  Base Class for RSA Signature Paddings (because they all share the hash function attribute)
  */
-class PKCSEncryptionPadding {
+class RSASignaturePadding {
 public:
-    /**
-     * @brief Get maximum data size that can encrypted using the PKCS padding
-     * @param key RSA public key that will be used for encryption
-     * @return the maximum size of the data that can be encrypted in bytes.
-     */
-    int getDataBlockSize(const AsymmetricPublicKey &key) const;
-
     /**
      * @brief Prepares the given openssl context with the padding specific parameters.
      * @param ctx OpenSSL context
+     * @param hashFunction The hash function to be used
      */
-    void prepareOpenSSLContext(openssl::SSL_EVP_PKEY_CTX_Ptr& ctx) const;
-
-private:
-    /**
-     * @brief Size overhead added by the PKCS padding on the RSA encryption
-     */
-    static const int c_pkcsMaxSizeSOverhead = 11;
+    virtual void prepareOpenSSLContext(openssl::SSL_EVP_PKEY_CTX_Ptr& ctx,
+                                       openssl::DigestTypes hashFunction) const = 0;
 };
 
+
 /**
- *  @brief RSASignaturePadding
+ * @brief NoPadding
  *
- *  Base Class for RSA Signature Paddings (because they all share the hash function attribute)
+ * This class defines the parameters specific to the RSA no padding mode.
  */
-class RSASignaturePadding  {
+class NoPadding : public RSAEncryptionPadding {
 public:
-    /**
-     * @brief Construtor for RSASignaturePadding
-     * @param hashFunction Hash function to be used
-     */
-    RSASignaturePadding(openssl::DigestTypes hashFunction);
-
-     /**
-      * @brief Get the hash function
-      *
-      * Getter method for the hash function.
-      *
-      * @return The hash function
-      */
-    openssl::DigestTypes getHashFunction() const;
-
-protected:
-    /**
-     * @brief Hash function to be used for signing
-     */
-    openssl::DigestTypes _hashFunction;
+    bool checkMessageSize(const AsymmetricPublicKey &key, size_t messageSize) const override;
```

Comment:
Well, lets assume there are three files
test.h:
```
class A {
public:
    virtual void func() = 0;
};

class B : public A {
// No declaration of func
};
```
test.cpp:
```
#include "test.h"

void B::func() {};
```
main.cpp:
```
#include "test.h"

int main () {
    A *a = new B();
    return 0;
}
```
That is what you are suggesting, right? At least for me this doesn't compile:
```
test.cpp:3:6: error: no declaration matches ‘void B::func()’
    3 | void B::func() {
      |      ^
test.cpp:3:6: note: no functions named ‘void B::func()’
In file included from test.cpp:1:
test.h:6:7: note: ‘class B’ defined here
    6 | class B : public A {
      |       ^
main.cpp: In function ‘int main()’:
main.cpp:4:18: error: invalid new-expression of abstract class type ‘B’
    4 |     A *a = new B();
      |                  ^
In file included from main.cpp:1:
test.h:6:7: note:   because the following virtual functions are pure within ‘B’:
    6 | class B : public A {
      |       ^
test.h:3:18: note: 	‘virtual void A::func()’
    3 |     virtual void func() = 0;
      |  
```
If you add the declaration though the code compiles just fine. Which means to me that `func()` has to be declared in `B` in order to implement it. Or am I missing something here?

Reply:
imho, if you define one, you need to declare one